pkg/helm/charts_test.go

```diff
@@ -0,0 +1,301 @@
+package helm_test
+
+import (
+	"context"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"time"
+
+	helmv2beta1 "github.com/fluxcd/helm-controller/api/v2beta1"
+	fluxmeta "github.com/fluxcd/pkg/apis/meta"
+	sourcev1beta1 "github.com/fluxcd/source-controller/api/v1beta1"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	pb "github.com/weaveworks/weave-gitops/pkg/api/profiles"
+	"github.com/weaveworks/weave-gitops/pkg/helm"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/repo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	testSecretName = "https-credentials"
+)
+
+var _ = Describe("RepoManager", func() {
+	Context("GetCharts", func() {
+		var repoManager *helm.RepoManager
+
+		BeforeEach(func() {
+			repoManager = &helm.RepoManager{}
+		})
+
+		It("returns all profiles in the repository", func() {
+			testServer := httptest.NewServer(http.FileServer(http.Dir("testdata/with_profiles")))
+			profiles, err := repoManager.GetCharts(context.TODO(), makeTestHelmRepository(testServer.URL), helm.Profiles)
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(profiles).To(ConsistOf(&pb.Profile{
+				Name:        "demo-profile",
+				Home:        "https://helm.sh/helm",
+				Sources:     []string{"https://github.com/helm/charts"},
+				Description: "Simple demo profile",
+				Keywords:    []string{"gitops", "demo"},
+				Maintainers: []*pb.Maintainer{
+					{
+						Name:  "WeaveWorks",
+						Email: "maintainers@weave.works",
+						Url:   "",
+					},
+					{
+						Name:  "CNCF",
+						Email: "",
+						Url:   "cncf.io",
+					},
+				},
+				Icon: "https://helm.sh/icon.png",
+				AvailableVersions: []string{
+					"1.1.0",
+					"1.1.2",
+				},
+			}))
+		})
+
+		When("no charts exist with the profiles tag", func() {
+			It("returns an empty list", func() {
+				testServer := httptest.NewServer(http.FileServer(http.Dir("testdata/no_profiles")))
+				profiles, err := repoManager.GetCharts(context.TODO(), makeTestHelmRepository(testServer.URL), helm.Profiles)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(profiles).To(BeEmpty())
+			})
+		})
+
+		When("server isn't a valid helm repository", func() {
+			It("errors", func() {
+				testServer := httptest.NewServer(http.FileServer(http.Dir("testdata")))
+				_, err := repoManager.GetCharts(context.TODO(), makeTestHelmRepository(testServer.URL), helm.Profiles)
+				Expect(err).To(MatchError(ContainSubstring("fetching profiles from HelmRepository testing/test-ns")))
+				Expect(err).To(MatchError(ContainSubstring("404")))
```

Comment:
Checking for a substring here might be a little problematic. Can we instead check the status code of the response?

Reply:
refactored the error message testing to be more explicit